[김경린] sprint10

k01zero opened the pull request:
# 요구사항

---

## 기본 요구사항

### **CSRF 보호 설정 리팩토링**

- [x]  CSRF 토큰 저장소를 쿠키 기반 저장 방식으로 리팩토링하세요.
- [x]  디스코드잇 프론트엔드에서 토큰을 헤더에 포함할 수 있도록 쿠키의 HTTP Only 속성은 `false`로 설정하세요.
    - 쿠키 이름: `XSRF-TOKEN`
    - 헤더 이름: `X-XSRF-TOKEN`

### **JWT 다루기**

- [x]  적절한 JWT 라이브러리를 추가하세요.
- [x]  다음의 요구사항을 만족하는 `JwtSession`을 구현하세요.
    - 패키지명: `com.sprint.mission.discodeit.security.jwt`
    - 다음의 정보를 포함하세요.
        - 사용자 정보
        - 발급된 엑세스 토큰
        - 발급된 리프레시 토큰
    - 데이터베이스를 통해 관리할 수 있습니다.
- [x]  다음의 요구사항을 만족하는 `JwtService`를 구현하세요.
    - 패키지명: `com.sprint.mission.discodeit.security.jwt`
    - [x] `UserDto` 정보로 토큰을 생성할 수 있습니다.
        - `JwtSession`을 같이 저장하세요.
    - [x] 토큰의 유효성을 검사할 수 있습니다.
    - [x] 리프레시 토큰을 활용해 엑세스 토큰을 재발급할 수 있습니다.
        - 리프레시 토큰 Rotation 전략을 활용합니다.
        - `JwtSession`을 같이 수정하세요.
    - [x] 리프레시 토큰을 무효화할 수 있습니다.
        - `JwtSession`을 같이 삭제하세요.
    - [x] 페이로드에는 다음의 정보를 포함합니다.
        - `iat`: 토큰 발행 시간
        - `exp`: 토큰 만료 시간
        - `userDto`: UserDto 정보 // userDto 의 전체 정보를 담길 원하는건지 아니면 일부도 괜찮은건지
       

### **로그인 리팩토링**

- [X]  로그인 API 스펙을 다음과 같이 변경합니다.
   
    | **엔드포인트** | **요청** | **응답** |
    | --- | --- | --- |
    | **`POST /api/auth/login`** | **`Body LoginRequest`** | **`200 String`, `401 ErrorResponse`** |
    
   - 엔드포인트와 요청은 이전과 동일합니다.
   - 응답 Body에 문자열로 엑세스 토큰을 포함하세요.
   - Cookie에 리프레시 토큰을 저장하세요.
    
- [x]  인증이 성공한 경우 `JwtService`를 활용해 토큰을 발급하세요.

### **엑세스 토큰 인증 필터 구현**

- 요청 헤더에 포함된 엑세스 토큰을 통해 인증하는 필터를 구현합니다.
- [x]  엑세스 토큰 인증이 필요한 요청에 대해서만 인증을 시도합니다.
- [x]  `JwtService`를 통해 토큰의 유효성을 검사합니다.
- [x]  유효한 토큰인 경우 인증 완료 처리합니다. 
- [x]  유효하지 않은 토큰인 경우 401 응답을 반환합니다. -> AuthenticationException

### **리프레시 토큰을 활용한 엑세스 토큰 정보 조회**

디스코드잇 프론트엔드에서 Access Token 정보는 브라우저 메모리에서만 관리합니다. 따라서 새로 고침 시 Access Token 정보가 날아가게 됩니다.

- [x]  Me API 스펙을 다음과 같이 변경합니다.

    | **엔드포인트** | **요청** | **응답** |
    | --- | --- | --- |
    | **`GET /api/auth/me`** | **`Header(자동 포함) Cookie: refresh_token=…`** | **`200 String`** |
    
   - 엔드포인트는 이전과 동일합니다.
   - 요청 헤더의 쿠키는 세션 대신 리프레시 토큰을 활용합니다.
   - 응답 Body에 문자열로 엑세스 토큰을 포함하세요.

디스코드잇 프론트엔드에서 엑세스 토큰을 파싱해 `UserDto`를 추출하여 사용합니다.

- [x] `JwtSession` 정보를 통해 해당 리프레스 토큰으로 발급된 엑세스 토큰을 반환합니다.
    - 엑세스 토큰을 새로 발급하지 않습니다.

### **로그아웃 리팩토링**

- [X]  로그아웃 API 스펙은 다음과 같이 변경합니다.
    
    | **엔드포인트** | **요청** | **응답** |
    | --- | --- | --- |
    | **`POST /api/auth/logout`** | **`Header(자동 포함) Cookie: refresh_token=…`** | **`200 Void`** |
    
   - 엔드포인트와 응답은 이전과 동일합니다.
   - 요청 헤더의 쿠키는 세션 대신 리프레시 토큰을 활용합니다.
    
- [X]  `JwtService`를 활용해 토큰을 무효화하세요.
- [X]  쿠키에 저장된 리프레시 토큰을 무효화하세요.

### **리프레시 토큰을 활용한 엑세스 토큰 재발급**

- [X]  토큰 재발급 API 스펙은 다음과 같습니다.
    
    | **엔드포인트** | **요청** | **응답** |
    | --- | --- | --- |
    | **`POST /api/auth/refresh`** | **`Header(자동 포함) Cookie: refresh_token=…`** | **`200 String`, `401 ErrorResponse`** |
   
   - 응답 Body에 문자열로 엑세스 토큰을 포함하세요.
   - Cookie에 리프레시 토큰을 저장하세요.
    
- [X]  요청 리프레시 토큰이 유효하고 `JwtSession`이 존재할 때 `JwtService`를 활용해 엑세스 토큰을 재발급합니다.
- [X]  리프레시 토큰이 유효하지 않거나, `JwtSession`이 존재하지 않으면 401 응답을 반환합니다.

### **권한 수정 리팩토링**

- 권한 수정 API 스펙인 이전과 동일합니다.
- [x]  권한이 수정된 사용자가 로그인 상태라면, 강제 로그아웃 되도록 합니다.
    - `JwtSession`을 활용하세요.
    
    강제 로그아웃 하더라도 해당 유저가 새로고침하지 않는다면,
    엑세스 토큰의 유효기간 동안은 이전의 권한으로 여전히 API 요청을 할 수 있습니다.
    이 버그에 대해서는 심화 요구사항을 수행하면서 해결합니다.
    
   ---

## 심화 요구사항

### **블랙리스트를 활용한 강제 로그아웃 고도화** [ ]

### **사용자 로그인 상태 리팩토링** [ ]

### **동시 로그인 제한 리팩토링** [ ]

---

## 주요 변경사항

---

## 스크린샷

---

## 멘토에게

? JwtSession 을 어떻게 받아들이면 될까요?
"다음의 요구사항을 만족하는 `JwtSession`을 구현"

라는 조건을 해석하는데 시간이 많이 걸렸는데,
accessToken 을 String 으로 저장하면 Stateless 한 장점이 상쇄되지 않을까요?
DB에 userId, accessToken, refreshToken 데이터를 테이블로 저장하고 해당 테이블을 통해서 accessToken, refreshToken 재발급 여부를 결정하는 게 드물까요?

강사님이 RefreshToken를 DB로 관리하는 코드를 제시해주셨어서, 요구 조건과 같은 형태가 실무에서도 자주 이용되는지 궁금합니다!

또한 이전 Session 관련 코드는 일부 지우고 시작했는데, 이렇게 하는 게 아니라 동시에 진행해야 했던 건지도 궁금합니다.

-

? 토큰의 무효화
revoked, replacedBy와 같은 기능을 쓰고 있는지라 삭제를 논리 삭제로 했습니다.
토큰 무효화시 JwtSession의 삭제, 이런 문구가 적혀있으니 논리 삭제가 아닌 물리 삭제가 더 요구 사항을 충족할 수 있는 방법이었을까요?

-

? 로그를 찍고 넘어갈 때와 에러를 던져야할 때
UserDto를 파싱에 실패했을 때 실제 User를 가져올 수 있도록 조회하도록 한 부분이 있습니다.
이때, 제 생각에는 매번 에러가 던져지면 애플리케이션이 돌아가는데 문제가 있을 것 같다는 생각이 들어
log.error("DB 에서 유저를 찾지 못했습니다. 토큰을 삭제합니다."); 이런 식으로 error 로 에러 로그를 찍어두고 반환은 Optional.empty()를 하도록 해놨는데.

문득 이런 판단에 대한 기준은 어떻게 해야하는 건지 궁금해졌습니다.
혹시 참고할 수 있는 문서나 생각의 기준이 있을까요?

k01zero committed:
rebase

k01zero committed:
rebase

k01zero committed:
second commit

k01zero committed:
Fix: BaseEntity.java의파라미터 변수명을 소ë¬

k01zero committed:
Fix: BaseEntity. BaseEntity.java의 'id'와 'createdAt'을 final로 지정

k01zero committed:
Refactor: 입출력 로직을 io 패키지로 분리

k01zero committed:
Test: ãJCFUserService 단위 테스트 추가

k01zero committed:
Test: JCFChannelService, JCFMessageService 단위 테스트 추가

k01zero committed:
refactor : Service layer상에 존재하던 데이터 처리 분리를 위한 repository 패키지 생성

k01zero committed:
refactor : UserRepository와 구현체 InMemoryUserRepository 작성

k01zero committed:
refactor : UserService 데이터 처리 부분을 Repository로 분리,  메서드 명및 메서드 내부 변경으로 인한 테스트 코드 수정

k01zero committed:
refactor : ChannelService 데이터 처리 부분을 Repository로 분리,  메서드 명및 메서드 내부 변경으로 인한 테스트 코드 수정

k01zero committed:
refactor : Message Service 데이터 처리 부분을 Repository로 분리,  메서드 명및 메서드 내부 변경으로 인한 테스트 코드 수정

k01zero committed:
refactor : Channel 모델ë에서 'user'ãã변 수ã수명로 변경í

k01zero committed:
fix : User, Channel, Message 모델의 생성자에 예외 처리ë 추가해 null과 공백 닉네임을 방지

k01zero committed:
refactor : ìã 서비스 로직의 인터페이스에서 추ìabstract 생략

k01zero committed:
bug fix

k01zero committed:
sprint mission 2 save

k01zero committed:
sprint mission 2 save

k01zero committed:
refactor : BaseEntity 모델에서ãㅇ ã u ì 지정서드명 변경

k01zero committed:
refactor : getById 메메서드 메서드 반환형을 Optional 래퍼 클래스로 지정한 후 이용된 코드들 orElse, orElseGet, ifPresentë¥등으로 수정

k01zero committed:
refactor : BaseEntityãë ㅠBaseEntity 모델에ì,efreshUpdatedAt 메서드에서 시간을 ë매개변수로 받지 않고 내내부에서 변경하도록 변경

k01zero committed:
refactor : getìrepository 인터페이스에서 getAll*  반환 자료형을 Map에서 Collection<*>으로 변경

k01zero committed:
refactor : Service dã인터페이스 showAll*의 반환 자료형을 Collection<*>으로 수정, êµ구íê´ë메서드를 구현한 구현체 수정, 단위 테스트 코드 수정

k01zero committed:
sprint-misiion-2-done

k01zero committed:
refactor(mentor) : File*Repository deleteUserById에 iãfë 삭제 여부에 따라서 ì¶문구 출력

k01zero committed:
Filrefactor(mento): File*Repository 생ì폴더 주ìì 주소의 구분자를OS체제를 고려하여 Paths.get().toString() 으로 변êã경, ìFileUserRepository와 FileChannelRepository 생성자ì ì수정

k01zero committed:
1-sprint-mission-3

k01zero committed:
1-sprint-mission-3

k01zero committed:
1-sprint-mission-4

k01zero committed:
part1-김경린-sprint5

k01zero committed:
part1-김경린-sprint5

k01zero committed:
part1-김경린-sprint5

k01zero committed:
part2-김경린-sprint6

k01zero committed:
part2-김경린-sprint6-modified

k01zero committed:
chore: 환경별 프로파일 설정 추가
- application.yml
- application-dev.yml
- application-prod.yml

k01zero committed:
feat: 로그 관리 설정 및 로깅 적용
- @Slf4j 어노테이션 추가
- application.yaml에 로깅 레벨 설정 (개발: DEBUG, 운영: INFO)
- logback-spring.xml 구성 (콘솔 & 파일 출력, 롤링 설정)
- 주요 서비스 및 컨트롤러 메서드에 로깅 추가

k01zero committed:
feat: 예외 처리 고도화
- ErrorCOde Enum 추가하여 예외 코드 및 메시지 정의
- DiscodeitException 및 도메인별 예외 계층 구조 설계
- 기존 예외를 커스텀 예외로 대체 (NoSuchElementException, IllegalArgumentException 등)
- ErrorResponse 및 @RestControllerAdvice 기반 글로벌 예외 핸들러 구현

k01zero committed:
feat: Spring Validation 추가 및 유효성 검사 구현
- 주요 Request DTO에 제약 조건 어노테이션 추가
- @Valid를 사용해 컨트롤러에서 요청 데이터 검증
- 유효성 검증 실패 시 MethodArgumentNotValidException 전역 예외 처리 및 ErrorResponse 반환

k01zero committed:
feat: Spring Boot Actuator 설정 추가 및 엔드포인트 구성

k01zero committed:
test: 서비스, 레포지토리, 컨트롤러 및 통합 테스트 추가

k01zero committed:
part2-김경린-sprint7

k01zero committed:
feat: Swagger 문서화 API 추가
- 컨트롤러 요청 파라미터 일부 수정

k01zero committed:
chore: docker 이미지 빌드 및 멀티 컨테이너 실행 구성
- Dockerfile 기반 애플리케이션 이미지 설정
- docker-compose.yml 작성 (애플리케이션 + PostgreSQL)
- 볼륨 설정 및 schema.sql 자동 실행 구성
- 볼륨 지정 과정에서 LocalBinaryContentStorage 관련 경로 설정 일부 수정

k01zero committed:
feat: AWS S3 기반 BinaryContentStorage 고도화
- S3 버킷 및 IAM 구성 기반 환경변수 설정 (.env)
- S3 저장소 테스트 클래스 추가
- S3BinaryContentStorage 구현 및 설정 분기 처리
- application.yaml 및 docker-compose 설정 수정

k01zero committed:
chore(docker): 멀티스테이지 빌드로 이미지 용량 최적화, 캐시 전략 개선

k01zero committed:
test GitHub Actions trigger

k01zero committed:
test GitHub Actions trigger in part2-김경린

k01zero committed:
Fix: 기존 테스트 수정 및 검증

- 테스트 코드에서 발생한 문제 해결 및 예외 처리 보강

k01zero committed:
FIx: Dockerfile 에 gradlew 실행 권한 부여하는 명령어 작성

k01zero committed:
Fix: deploy.yml 워크플로우에 cluster, service 환경 변수 이름 GitHub Actions 에 Variables 로 지정한 이름과 일치하도록 수정

k01zero committed:
refactor: 환경 설정 개선 및 예외 처리 리팩토링

- application.yml | active : DB 환경변수 지정, info : 환경 변수로 DB 표시
- application-dev.yml | H2 MODE=PostgreSQL 설정 추가
- GlobalExceptionHandler 보완(전체 예외 처리 관련) 및 ErrorCode 추가(입력값 유효성 검증 관련)
- mockito 의존성 제거

k01zero committed:
Fix: 테스트 코드 수정

k01zero committed:
Delete .env

PostgreSQL DB 설정이 들어있던 초기 env 파일

k01zero committed:
sprint8

k01zero committed:
sprint8

k01zero committed:
refactor: S3Presigner 필드 선언및 생성자 초기화

k01zero committed:
fix: Dockerfile, CMD 명령어에 "exec" 추가
- 컨테이너에서 시그널 전달이 java 로 안 되는 문제 해결

k01zero committed:
refactor: settings.gradle, build.gradle copy 순서 변경

k01zero committed:
fix: Dockerfile wrapper 을 통한 빌드로 변경
- gradle 빌드와 wrapper 빌드 혼용되던 부분 수정
- RUN gradle 제거
- gradlew 관련 파일 copy 추가

k01zero committed:
feat: Spring Security 환경 설정
- SecurityConfig 설정 : SecurityFilterChain Bean 등록시 LoginFilter 제외
- application.yml : Spring Security 모듈의 로깅 레벨 trace 로 설정

k01zero committed:
feat: CSRF 보호 설정하기
- CSRF 토큰 발급하는 API 구현
- SecurityFilterChain 옵션에서 해당 API 엔드포인트 필터링 제외

k01zero committed:
refactor: 프론트엔드 코드 교체

k01zero committed:
refactor: 프론트엔드 코드 교체

k01zero committed:
feat: 회원가입 고도화
- 비밀번호 해싱 BCryptPasswordEncoder

k01zero committed:
feat: 기본 인증 구현
- 기존 로그인 관련 코드 제거
- 로그인 API Controller 구현
- SecurityConfig 커스텀 필터 등록및 사용하지 않는 필터(formLogin)는 disable으로 설정
- CustomLoginFilter : JSON 요청을 받는 로그인 커스텀 필터 구현
- CustomUserDetails*.java : 커스텀 필터의 User 검증용 인터페이스 구현체
- Login*Handler : 커스텀 필터의 응답(Success/Failure) 구현체

k01zero committed:
feat: 세션을 활용한 현재 사용자 정보 조회
- 세션 사용자 조회 API 구현
- 유저 인증관련 에러 추가(ErrorCode, SecurityException, UserUnauthorizedException)
- SecurityConfig : h2-console URL 관련 인증 설정 추가
- SecurityConfig : .anyRequest().authenticated() 추가

k01zero committed:
feat: 로그아웃 기능 구현
- 로그아웃에 대한 CSRF 필터 다운
- 로그아웃시 세션 무효화 처리
- 로그아웃시 SecurityContext 초기화

k01zero committed:
feat: 권한 관리 기능 구현
- 회원 가입 시 모든 사용자 user 권한 소유
- AdminInitializer : admin 계정 초기화
- 사용자 권한 API 스펙 구현

k01zero committed:
feat: 인가 처리 추가
- 그외 : AdminInitializer 에서 admin 계정 생성시 userStatus 누락 추가

k01zero committed:
feat: 인가 고도화
- 사용자 정보 수정, 삭제: 본인 또는 ROLE_ADMIN
- 메시지 수정: 메시지 작성자 본인
- 메시지 삭제: 메시지 작성자 본인 또는 ROLE_ADMIN
- 읽음 상태 생성, 수정: 본인

k01zero committed:
feat: 세션 고정 보호 설정

k01zero committed:
feat: 동시 로그인 제한

k01zero committed:
feat: Remember-Me 자동 로그인 기능 추가

k01zero committed:
refactor: UserStatus 엔티티 관련 모드 코드 삭제

k01zero committed:
feat: 사용자 로그인 상태 고도화
- Session 정보를 활용해 사용자의 로그인 상태를 판단
- UserMapper : Session 이 활성화되어 있고, Session에 포함된 Authentication 이 유효할시 "온라인" 상태로 판단하여 Dto 로 반환

k01zero committed:
refactor: 정적 리소스 접근 권한 허용

k01zero committed:
refactor: 정적 리소스 교체

k01zero committed:
fix: CustomLoginFilter 에 세션 인증 전략 적용
- Session 의 SessionRegistry 등록 누락으로 동시 세션 제어 등의 기능이 처리되지 않았음
- CustomLoginFilter : SessionAuthenticationStrategy 오버라이드
- SecurityConfig : SessionAuthenticationStrategy 빈 등록 및 CustomLoginFilter 빈 생성시 적용

k01zero committed:
refactor: yaml 파일 수정
- 서블릿 세션 쿠키 설정 추가

k01zero committed:
chore: version 2.2-M10 으로 변경

k01zero committed:
feat: CSRF 보호 설정 리팩토링
- SecurityConfig : CookieCsrfTokenRepository 빈 설정
- SecurityConfig : HttpSecurity 의 csrf 설정에 빈으로 설정한 저장소 지정

k01zero committed:
feat: JWT 다루기
- JWT 라이브러리 추가
- JwtSession 엔티티 구현
- JwtService 기능 구현(토큰 유효성 검사, AccessToken 재발급, RefreshToken 무효화, Jwt 생성시 iat, exp, userDto 포함)

k01zero committed:
feat: 로그인 리팩토링
- SecurityConfig : jwt 설정에 맞게 마이그레이션
- 로그인 api 기능 구현
- UserLoginRequest > LoginRequest 이름 변경과 변경에 따른 리팩토링

k01zero committed:
feat: 엑세스 토큰 인증 필터 구현
- JwtAuthenticationFilter 필터 구현
- SecurityConfig : 필터 추가

k01zero committed:
feat: 리프레시 토큰을 활용한 엑세스 토큰 정보 조회
- me API 구현
- JwtService : Refresh Token 으로 Access Token 을 찾는 로직 추가

k01zero committed:
feat: 로그아웃 리펙토링
- 로그아웃 API 구현
- DB, Cookie 에 포함된 토큰 무효화

k01zero committed:
feat: 리프레시 토큰을 활용한 엑세스 토큰 재발급
- JwtService 의 재발급 메서드에 jwt 토큰 유효성 검증 추가
- 에러 처리 추가

k01zero committed:
refactor: 권한 수정 리팩토링
- SessionRepository 를 통해 세션으로 무효화하던 작업을 JwtSessionRepository 을 통해 무효화하는 것으로 변경

k01zero committed:
refactor: 서명 키(Signgin Key) 한 번만 생성하도록 변경
- getSigningKey() 메서드 호출마다 SecretKey 객체를 생성하여 반환하던 부분 변경
- 서명 키 필드를 추가하여 없을 시에만 생성, 있을 시 재활용

k01zero committed:
refactor: 프론트엔드 정적 리소스 교체

k01zero committed:
fix: findAllByUserIdAndRevokedFalse 네이밍 메서드 오타 수정
- JPA Repository 네이밍 규칙 준수

k01zero committed:
fix: csrf-token URL 접근 권한 허용

k01zero committed:
fix: csrfFilter 필터에서 일어나는 403 에러 오류 해결
- XorCsrfTokenRequestAttributeHandler 대신 CsrfTokenRequestAttributeHandler (XOR 기능이 없는 기본 핸들러)로 교체

k01zero committed:
refactor: Spring security 관련 코드 전반을  base 11 코드로 교체

yeongunheo commented on src/main/java/com/sprint/mission/discodeit/security/jwt/JwtService.java:
```diff
@@ -0,0 +1,324 @@
+
+  // JwtTokenProvider 역할
+  private String generateToken(UserDto userDto, LocalDateTime issuedAtLdt,
+      LocalDateTime expiresAtLdt, String tokenType) {
```
tokenType은 따로 ENUM으로 만들어 관리해도 좋을 것 같아요.

yeongunheo commented on src/main/java/com/sprint/mission/discodeit/config/SecurityConfig.java:
```diff
@@ -0,0 +1,225 @@
+        .httpBasic(AbstractHttpConfigurer::disable)
+        // 세션 관리 설정
+        .sessionManagement(
+            session -> session.sessionCreationPolicy(SessionCreationPolicy.STATELESS))
```
👍

yeongunheo commented on src/main/java/com/sprint/mission/discodeit/config/SecurityConfig.java:
```diff
@@ -0,0 +1,225 @@
+                    "/favicon.ico",
+                    "/h2-console/**",
+                    "/api/auth/login",
+//                "/api/auth/csrf-token",
```
이 주석 해제해야 csrf 요청 시 403 에러가 발생하지 않을 것 같아요.

yeongunheo reviewed (approved):
경린님, 안녕하세요.  

기본 요구사항 잘 구현해주셨네요 💯   

몇가지 참고할만한 코멘트만 남겨두었습니다!  

질문도 같이 남겨주셔서 답변은 아래에 남겨둘게요.   

첫번째 질문은 멘토링 시간이 이야기해보는게 더 좋을 것 같아서 나머지 2개 질문만 제 생각 남겨두겠습니다!

--- 

> Q2. 토큰의 무효화
revoked, replacedBy와 같은 기능을 쓰고 있는지라 삭제를 논리 삭제로 했습니다.
토큰 무효화시 JwtSession의 삭제, 이런 문구가 적혀있으니 논리 삭제가 아닌 물리 삭제가 더 요구 사항을 충족할 수 있는 방법이었을까요?  

저라면 지금처럼 간단한 시스템은 물리삭제로 구현했을 것 같아요.  

JwtSession이 DB에 계속 쌓이게되면 매번 인증할 때마다 테이블을 조회할 때 성능이 느려질 수 있거든요.  

다만, 나중에 대규모 시스템에서 보안이 중요해질 때는 물리삭제보다는 논리삭제가 더 좋을 것 같습니다. 나중에 인증에 문제가 생겼을 떄 액세스토큰/리프레시토큰을 추적해야하는데 이때 논리삭제로 DB에 데이터가 남아있는게 디버깅이 더 용이하거든요.  

논리삭제, 물리삭제는 각각 장단점이 명확해서 이를 정리해봐도 좋을 것 같아요.



> Q3. 로그를 찍고 넘어갈 때와 에러를 던져야할 때
UserDto를 파싱에 실패했을 때 실제 User를 가져올 수 있도록 조회하도록 한 부분이 있습니다.
이때, 제 생각에는 매번 에러가 던져지면 애플리케이션이 돌아가는데 문제가 있을 것 같다는 생각이 들어
log.error("DB 에서 유저를 찾지 못했습니다. 토큰을 삭제합니다."); 이런 식으로 error 로 에러 로그를 찍어두고 반환은 Optional.empty()를 하도록 해놨는데.

단순히 로그를 남기는것 vs 에러를 던지는 것 중에 어떤걸 선택할 지는 맥락에 따라 다른데요,  

제 생각엔 로직 상 심각한 문제이며 에러를 던져서 메서드를 호출하는 쪽에 에러를 뱉어야할 경우는 에러를 던지는게 더 좋다고 생각해요.  

다시 말해, 에러를 뱉어야 하는 상황에서 단순히 로그를 남긴다면 이것은 에러 추적을 어렵게 만들게 되겠죠. 나중에 더 큰 에러로 이어질 가능성이 높습니다.

경린님이 작성하신 코드 보면 대표적으로 아래 2가지 케이스에서 에러 로그를 남기고 Optional.empty()를 반환하고 있는데요,
- Claims 객체에서 userDto 클레임 파싱하여 UserDto 객체로 변환할때 역직렬화 실패
- oldSession의 userId로 DB에서 유저 정보 조회시 실패

로그 레벨로 error로 남기고 있어서 이 두 로직에 있어서는 경린님 방식이 괜찮다고 생각해요.

k01zero committed:
feat: 비동기 적용하기
- 비동기 설정및 재시도 설정
- 재시도 횟수 초과 후 복구 로직 추가
- 실패 정보 AsyncTaskFailure 추가

k01zero committed:
fix: (누락) 메세지 생성시 저장소 업로드 기능 추가

k01zero committed:
fix: 비동기 작동 오류 해결
- 메서드 시그니처 변경 : 반환값 UUID -> CompletableFuture<UUID>

k01zero committed:
feat: 비동기 적용하기 : 비동기 처리 성공/실패에 따라 BinaryContent 의 상태를 업데이트

k01zero committed:
fix: TransactionRequiredException 오류 해결
- TransactionTemplate 을 통해 트랜잭션 관리

k01zero committed:
feat: 비동기 적용하기 - 동기와 비동기 성능 차이 비교

k01zero committed:
feat: 알림 기능 추가하기
- 알림 관련 API 구현
- ReadStatus 엔티티 리팩토링
- 비동기 방식 + 이벤트 리스너로 알림 생성

k01zero committed:
refactor: Spring Security 관련 코드 재교체

k01zero committed:
refactor: 교체한 Spring Security 코드 관련 리팩토링

k01zero committed:
refactor: 정적 리소스 교체

k01zero committed:
feat: 캐시 적용하기
- Spring Cache, Caffeine Cache 이용을 위한 셋업
- 사용자별 채널 목록 조회, 사용자별 알림 목록 조회, 사용자 목록 조회에 로컬 캐시 적용
